Add section about metadata (#944)
jamulussoftware/jamulussoftware.github.io: _posts/2023-07-30-Server-Metadata.md

---
layout: post

> title: "Adding metadata to Servers"

Reviewer: ```suggestion
title: "Allow or block bots"
```

I hope the title convention in this file supports _sentence casing_ because that's my jam. I also know you often use _-ing_ convention, but this does not seem to be the title convention in this file, thank goodness.

If you want to introduce the generalized architecture, I won't stand in your way. The first goal is to make sure everyone who wants to set a bot policy can get it done. This includes total noob operators, who don't know English. That's why I suggest just covering the one bot convention in this file, in a much shorter article. I would volunteer, except I believe there are adequate bot controls now. We should brainstorm additional metadata that might also make sense to introduce, but I'm not aware of anything yet.

Author: Ok. Then I'd accept this title for now, but the page should be updatable to cover more metadata.

lang: "en"
heading: "Metadata for servers"
---

> To provide information about Servers or clarify policies for Bots, server owners can add metadata to their servers' welcome message. This page gives a rough format about how the metadata could look like.

Reviewer: ```suggestion
To provide information about Servers and specify policies for bots, a server owner can add metadata to their server welcome message. This page describes metadata a welcome message might include.
```

Can't decide if _might_ makes sense. Want to communicate that it's not required, and it's not limited to these. I guess I wonder which are supported in any way. Admin contact, for example: Unless we surface this in some way, perhaps it's unwise to invite people to include it. Admin contact is the kind of data Jamulus Explorer might include. Until a tag is supported in some way, I wouldn't even want to mention it. So for example, I'd roll out support for a bot standard before rolling out this new page.

Reviewer: We are somewhat compelled here to define bots. While people run stationary bots on their own servers, this is more about bots that connect to public servers (_roaming bots_ or _docking bots_). I'm only aware of the two I'm running. I'd call one a _snippet bot_, and the other is a _streaming bot_, but I haven't thought of a word to describe both, except _listening bot_, which I find uncomfortable. At this time there isn't a recording bot; establishing this featureset will improve my ability to provide one. We haven't quite established a v1 featureset... maybe the first release can just be YES/NO, and future releases can specify roaming bot policy with greater granularity.

Author: I mean the site needs to be very general. 

> I guess I wonder which are supported in any way

Anything which fits into the metadata box in theory. But that's up to the people using the metadata. 

I wouldn't add a might - but I don't feel too strong about it.

> Unless we surface this in some way, perhaps it's unwise to invite people to include it.

People could still include their home address, coordinates or even social media profile in there. But yes, normally you'd add an email or website.

Author: > So for example, I'd roll out support for a bot standard before rolling out this new page.

Still we should have a standard somewhere - at least a very general one people can follow.

Author: > maybe the first release can just be YES/NO, and future releases can specify roaming bot policy with greater granularity.

Yes sure. Then we'd drop the thing with the bot policy all together and just talk about metadata in general?

Reviewer: I think now that metadata for bot policy is basically the only solid case for metadata. If you want to reach users, use clear text html. Bot policy reaches  bot provider probes.

<!--more-->

> You can add metadata to your server for many purposes. Adding contact information like the E-Mail or Website of the Server administrator helps users to contact the administrator if something goes wrong with the server. Also adding policies for bots allows them to judge if they can or cannot connect to the server.

Reviewer: It only helps if there's a straightforward way for users to find or use this information. Including this information in clear text is arguably much more useful than hiding it as our HTML will do. If I wanted users to easily contact me, providing a visible email or website address seems foolproof, while a hidden metadata entry is... useless at this point, right? So I'm not sure we've thought through how this guidance would help. Eventually it might be discoverable at both Jamulus Explorer and Live, but until then, I'm not sure it's worth recommending.

Author: Tools can extract it - and probably also jamulus explorer could parse it and display it in a new field.

Reviewer: The user's discovery is highly diminished vs. clear text. I don't see how we could beat guidance that says, "Hey, put your email address in your welcome message." Putting it in hidden text gains us... what?

Reviewer 2: The _point_ of "metadata" is _not_ to be human-readable.  It's directed at machines and should be in an easily _machine-parseable_ format.

Anyone can put any human-visible text in their welcome message already.

Using `<!--` and `-->` allows a parser to delimit the metadata - which, for a parser, is fairly necessary - otherwise it's guesswork.

The idea here was that we'd say, "Hey, if you want to tag up your server so it has machine readable data, here's how we suggest you do it."

Then if someone says "Hey, it'd be cool if you tag your server up with \<some info I want to parse> using \<this suggested tag>" we already have that idea embedded in the documentation and it's not alien to server operators.  It would be down to the person wanting to scrape the information to make sure they weren't clashing with anything that existed for other purposes.  It might be that we end up with a page logging metadata tags and their meanings, as people add them - but that wouldn't be _owned_ by anyone other than the people using the tags.

It's _definitely_ never going to be something Jamulus itself is interested in.

Reviewer: Why not? Require an admin contact email someday. You could announce new releases via email. It'd be grand.

Author: I think this goes against the policy of Jamulus being open. Also alerting people via E-Mail should be opt in.

> The point of "metadata" is not to be human-readable. It's directed at machines and should be in an easily machine-parseable format.

Exactly. I think we should drop the talk about admin contact here.
Focus on a generalized metadata format.

Reviewer: I've suggested focusing on bot policy, and also suggested brainstorming good candidates for metadata usage. Surely someone could think of more than one usage of this generalized format before you tout it.

## How to add metadata

Metadata should be added to the welcome message. If you are running a GUI server, paste the metadata into the welcome message box. Headless servers can use the `--welcomemessage` CLI argument to provide a path to a file containing the welcome message.

## Format of metadata

> In general, a metadata tag is a HTML comment which looks like this:

Reviewer: ```suggestion
A metadata tag is a HTML comment that looks like this:
```
